[SQL]Unsupported SQL types

Found more unmapped/unhandled Sql types:

- `Unsupported SQL type: 7`
- `Caused by: java.lang.NullPointerException: Cannot invoke "java.sql.Timestamp.toInstant()" because "value" is null`
```
Caused by: java.lang.NullPointerException: Cannot invoke "java.sql.Timestamp.toInstant()" because "value" is null
	at com.sneaksanddata.arcane.framework.services.mssql.MsSqlStreamingDataProvider.stream$$anonfun$1$$anonfun$1(MsSqlStreamingDataProvider.scala:38)
```

--- a/src/main/scala/services/mssql/MsSqlStreamingDataProvider.scala
+++ b/src/main/scala/services/mssql/MsSqlStreamingDataProvider.scala
@@ -35,8 +35,18 @@ class MsSqlStreamingDataProvider(dataProvider: MsSqlDataProvider,
       ZStream.unfoldZIO(None)(v => continueStream(v)).flatMap(readDataBatch)
     stream
       .map( row => row.map{
-        case DataCell(name, ArcaneType.TimestampType, value) => DataCell(name, ArcaneType.TimestampType, LocalDateTime.ofInstant(value.asInstanceOf[Timestamp].toInstant, ZoneOffset.UTC))
-        case DataCell(name, ArcaneType.ByteArrayType, value) => DataCell(name, ArcaneType.ByteArrayType, ByteBuffer.wrap(value.asInstanceOf[Array[Byte]]))
+        case DataCell(name, ArcaneType.TimestampType, value) if value != null
+          => DataCell(name, ArcaneType.TimestampType, LocalDateTime.ofInstant(value.asInstanceOf[Timestamp].toInstant, ZoneOffset.UTC))
+        
+        case DataCell(name, ArcaneType.TimestampType, value) if value == null
+          => DataCell(name, ArcaneType.TimestampType, null)
+        
+        case DataCell(name, ArcaneType.ByteArrayType, value) if value != null
+          => DataCell(name, ArcaneType.ByteArrayType, ByteBuffer.wrap(value.asInstanceOf[Array[Byte]]))
+
+        case DataCell(name, ArcaneType.ByteArrayType, value) if value == null
+          => DataCell(name, ArcaneType.ByteArrayType, null)
+        
         case other => other
       })
 

--- a/src/main/scala/utils/SqlUtils.scala
+++ b/src/main/scala/utils/SqlUtils.scala
@@ -57,7 +57,12 @@ object SqlUtils:
     case java.sql.Types.LONGNVARCHAR => Success(ArcaneType.StringType)
     case java.sql.Types.DOUBLE => Success(ArcaneType.DoubleType)
     case java.sql.Types.INTEGER => Success(ArcaneType.IntType)
+
     case java.sql.Types.FLOAT => Success(ArcaneType.FloatType)
+    // The ISO synonym for real is float(24).
+    // See: https://learn.microsoft.com/en-us/sql/t-sql/data-types/float-and-real-transact-sql?view=sql-server-ver16
+    case java.sql.Types.REAL => Success(ArcaneType.FloatType)
+    
     case java.sql.Types.SMALLINT => Success(ArcaneType.ShortType)
     case java.sql.Types.TIME => Success(ArcaneType.TimeType)
     case java.sql.Types.NCHAR => Success(ArcaneType.StringType)

--- a/src/test/scala/services/connectors/mssql/MsSqlConnectorsTests.scala
+++ b/src/test/scala/services/connectors/mssql/MsSqlConnectorsTests.scala
@@ -2,7 +2,7 @@ package com.sneaksanddata.arcane.framework
 package services.connectors.mssql
 
 import models.{ArcaneSchemaField, DataCell, Field, MergeKeyField}
-import models.ArcaneType.{BigDecimalType, ByteArrayType, IntType, LongType, StringType, TimestampType}
+import models.ArcaneType.{BigDecimalType, ByteArrayType, FloatType, IntType, LongType, StringType, TimestampType}
 import services.connectors.mssql.util.TestConnectionInfo
 import services.mssql.query.{LazyQueryResult, ScalarQueryResult}
 import services.mssql.{ColumnSummary, ConnectionOptions, MsSqlConnection, QueryProvider}
@@ -50,7 +50,7 @@ class MsSqlConnectorsTests extends flatspec.AsyncFlatSpec with Matchers:
         None), con)
 
   def createTable(tableName: String, con: Connection): Unit =
-    val query = s"use arcane; drop table if exists dbo.$tableName; create table dbo.$tableName (x int not null, y int, z DECIMAL(30, 6), a VARBINARY(MAX), b DATETIME, [c/d] int)"
+    val query = s"use arcane; drop table if exists dbo.$tableName; create table dbo.$tableName (x int not null, y int, z DECIMAL(30, 6), a VARBINARY(MAX), b DATETIME, [c/d] int, e real)"
     val statement = con.createStatement()
     statement.executeUpdate(query)
 
@@ -63,13 +63,13 @@ class MsSqlConnectorsTests extends flatspec.AsyncFlatSpec with Matchers:
   def insertData(con: Connection): Unit =
     val statement = con.createStatement()
     for i <- 1 to 10 do
-      val insertCmd = s"use arcane; insert into dbo.MsSqlConnectorsTests values($i, ${i+1}, null, CAST(123456 AS VARBINARY(MAX)), '2023-10-01 12:34:56', 0)"
+      val insertCmd = s"use arcane; insert into dbo.MsSqlConnectorsTests values($i, ${i+1}, null, CAST(123456 AS VARBINARY(MAX)), '2023-10-01 12:34:56', 0, 0)"
       statement.execute(insertCmd)
     statement.close()
 
     val updateStatement = con.createStatement()
     for i <- 1 to 10 do
-      val insertCmd = s"use arcane; insert into dbo.MsSqlConnectorsTests values(${i * 1000}, ${i * 1000 + 1}, ${i * 1000 + 2}, CAST(123456 AS VARBINARY(MAX)), '2023-10-01 12:34:56', 0)"
+      val insertCmd = s"use arcane; insert into dbo.MsSqlConnectorsTests values(${i * 1000}, ${i * 1000 + 1}, ${i * 1000 + 2}, CAST(123456 AS VARBINARY(MAX)), '2023-10-01 12:34:56', 0, 0)"
       updateStatement.execute(insertCmd)
 
   def deleteData(connection: Connection, primaryKeys: Seq[Int]): ZIO[Any, Throwable, Unit] = ZIO.scoped {
@@ -150,6 +150,7 @@ class MsSqlConnectorsTests extends flatspec.AsyncFlatSpec with Matchers:
         |tq.[a],
         |tq.[b],
         |tq.[c/d],
+        |tq.[e],
         |@currentVersion AS 'ChangeTrackingVersion',
         |lower(convert(nvarchar(128), HashBytes('SHA2_256', cast(tq.[x] as nvarchar(128))),2)) as [ARCANE_MERGE_KEY]
         |FROM [arcane].[dbo].[MsSqlConnectorsTests] tq""".stripMargin
@@ -178,6 +179,7 @@ class MsSqlConnectorsTests extends flatspec.AsyncFlatSpec with Matchers:
         |CAST(0 as BIGINT) as SYS_CHANGE_VERSION,
         |'I' as SYS_CHANGE_OPERATION,
         |tq.[y],
+        |tq.[e],
         |@currentVersion AS 'ChangeTrackingVersion',
         |lower(convert(nvarchar(128), HashBytes('SHA2_256', cast(tq.[x] as nvarchar(128))),2)) as [ARCANE_MERGE_KEY]
         |FROM [arcane].[dbo].[MsSqlConnectorsTests] tq""".stripMargin
@@ -234,6 +236,7 @@ class MsSqlConnectorsTests extends flatspec.AsyncFlatSpec with Matchers:
       Field("a", ByteArrayType),
       Field("b", TimestampType),
       Field("cd", IntType),
+      Field("e", FloatType),
       Field("ChangeTrackingVersion", LongType),
       MergeKeyField)
     Unsafe.unsafe(implicit unsafe => runtime.unsafe.runToFuture(connection.getSchema)) map { schema =>
@@ -259,7 +262,7 @@ class MsSqlConnectorsTests extends flatspec.AsyncFlatSpec with Matchers:
         result <- connection.backfill.runCollect
         head = result.head
     yield {
-      head should have length 10
+      head should have length 11
     }
     
     Unsafe.unsafe(implicit unsafe => runtime.unsafe.runToFuture(task))
@@ -348,6 +351,7 @@ class MsSqlConnectorsTests extends flatspec.AsyncFlatSpec with Matchers:
       "a",
       "b",
       "cd",
+      "e",
       "ChangeTrackingVersion",
       "ARCANE_MERGE_KEY",
     )

